### PR TITLE
catch JS exceptions when using InvokeAsync and throw them in .NET

### LINKED ISSUE #40272: InvokeAsync<IJSObjectReference?>() call does not return when js value is null or undefined
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When InvokeAsync<IJSObjectReference?>() is called and the invoked js function returns null or undefined, then an error is thrown in js and the execution of the caller stops.

### Expected Behavior

Either a null value is returned or an error with the necessary information is thrown on the .NET side.

### Steps To Reproduce

Create a new wasm project.
Add the following div element to Index.razor
```html
<div id="test123">I'm a test</div>
```
Add the following code block to Index.razor and inject a IJsRuntime instance
```c#
@code {
    protected async override Task OnAfterRenderAsync(bool firstRender)
    { 
        if (firstRender)
        {
            try
            {
                var testNotNull = await js.InvokeAsync<IJSObjectReference?>("FailWhenIsNull", false);
                System.Console.WriteLine("Success not null");
                var testNull = await js.InvokeAsync<IJSObjectReference?>("FailWhenIsNull", true);
                System.Console.WriteLine("Success null");
            }
            catch(JSException)
            {
                System.Console.WriteLine("Some failure");
            }
            finally{
                System.Console.WriteLine("Finished");
            }
            
        }
    }
}
```
Add the script to index.html
```js
<script>
        window.FailWhenIsNull = (isNull) => {
            if (isNull) {
                return null;
            } else {
                return document.getElementById('test123');
            }
        }
</script>
```
Start the application and look into the web console.
grafik

### Exceptions (if any)

_No response_

### .NET Version

6.0.100

### Anything else?

_No response_

### LINKED ISSUE #41934: JSRuntime "open" with "_blank" throws TaskCanceledException 60 seconds later
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Clicking a button that calls the following results in a TaskCanceledException about 60 seconds later.

```c#
await JSRuntime.InvokeAsync<string>("open", $"https://www.microsoft.com", "_blank");
```

### Expected Behavior

No TaskCanceledException is thrown

### Steps To Reproduce

1. Run the project (tested with `dotnet run` and VS Mac (ARM)) in Safari or Firefox on macOS
2. Click a button that calls the following: `await JSRuntime.InvokeAsync<string>("open", $"https://www.microsoft.com", "_blank");` via an `@onclick` method
3. Notice in the exception in the browser console (and in VS Mac (ARM) as well if started there)

### Exceptions (if any)

browser console error:
> Error: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)

VS Mac (ARM) error:
> Exception thrown: 'System.Threading.Tasks.TaskCanceledException' in System.Private.CoreLib.dll

### .NET Version

6.0.300

### Anything else?

.NET SDK (reflecting any global.json):
 Version:   6.0.300
 Commit:    8473146e7d

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  12.4
 OS Platform: Darwin
 RID:         osx.12-arm64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.300/

Host (useful for support):
  Version: 6.0.5
  Commit:  70ae3df4a6

.NET SDKs installed:
  6.0.101 [/usr/local/share/dotnet/sdk]
  6.0.200 [/usr/local/share/dotnet/sdk]
  6.0.201 [/usr/local/share/dotnet/sdk]
  6.0.202 [/usr/local/share/dotnet/sdk]
  6.0.203 [/usr/local/share/dotnet/sdk]
  6.0.300 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.2 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.2 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.4 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

### PR DESCRIPTION
**Pull request by campersau**
# catch JS exceptions when using InvokeAsync and throw them in .NET

Exceptions in `createJSCallResult` and `stringifyArgs` were only thrown in JS before and .NET never received them.

Examples of exceptions which are now reaching .NET:
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L591
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L167-L188
- https://github.com/dotnet/aspnetcore/blob/371bf375f5bf6e48ab6b33e24ad0c73a940b4ae4/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L148

Fixes #40272 (references the last exception from the list above)
Fixes #41934 (exception in `stringifyArgs`)

### CODE REVIEW

File: src/Components/WebAssembly/WebAssembly/test/JSObjectReferenceJsonConverterTest.cs [hunk 1]
```diff
@@ -17,6 +17,20 @@ public JSObjectReferenceJsonConverterTest()
         JsonSerializerOptions = DefaultWebAssemblyJSRuntime.Instance.JsonSerializerOptions;
     }
 
+    [Fact]
+    public void Read_ReadsJson_IJSObjectReference()
```
Comment (campersau) [inline src/Components/WebAssembly/WebAssembly/test/JSObjectReferenceJsonConverterTest.cs:21]: Unrelated bonus test.

**Comment by mkArtakMSFT** [comment]
Thanks for your PR, @campersau.
@MackinnonBuck can you please review this? Thanks!

**Comment by campersau** [comment]
Updated to also catch exceptions during `stringifyArgs` which is the case in https://github.com/dotnet/aspnetcore/issues/41934 and added tests for synchronus `Invoke` to show that the behavior is the same as `InvokeAsync`, both throwing exceptions.

**Review by MackinnonBuck — APPROVED** [review]
LGTM

**Comment by japui** [comment]
catch JS exceptions when using InvokeAsync and throw them in .NET

**Comment by MackinnonBuck** [comment]
Thanks @campersau!

**Comment by datvm** [comment]
Hi is this PR actually online yet? I am using .NET 7 Blazor WebAssembly (7.0.202) and I still have this issue. The code runs fine if JS returns something but does not returns when JS returns `undefined` (Error printed in Console but nothing happens on .NET side).

```cs
    public static async Task<T> GetGlobalObject<T>(this IJSRuntime js, string key) =>
        await js.InvokeAsync<T>(
            GetFullFuncName("getGlobalObj"),
            key);

    public static async Task<IJsWindowReference?> GetOpenerAsync(this IJSRuntime js)
    {
        var opener = await js.GetGlobalObject<IJSObjectReference?>("opener");
        return opener?.AsWindow(js);
    }
```

```js
    getGlobalObj(key) {
        return globalThis[key];
    }
```

**Pull request merged**